crates/jstzd/src/task/octez_node.rs

```diff
@@ -0,0 +1,240 @@
+use super::Task;
+use anyhow::Result;
+use async_dropper_simple::{AsyncDrop, AsyncDropper};
+use async_trait::async_trait;
+use std::{fs::File, path::PathBuf, sync::Arc};
+use tokio::sync::RwLock;
+
+#[cfg(test)]
+use tests::{Child, InnerOctezNode};
+
+#[cfg(not(test))]
+use octez::OctezNode as InnerOctezNode;
+#[cfg(not(test))]
+use std::process::Child;
+
+pub struct OctezNodeConfig {
+    binary_path: PathBuf,
+    data_dir: PathBuf,
+    network: String,
+    rpc_endpoint: String,
+    num_connections: u32,
+    log_file: PathBuf,
+    options: Vec<String>,
+}
+
+impl OctezNodeConfig {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        binary_path: &str,
+        data_dir: &str,
+        network: &str,
+        rpc_endpoint: &str,
+        num_connections: u32,
+        log_file: &str,
+        options: &[&str],
+    ) -> Self {
+        OctezNodeConfig {
+            binary_path: PathBuf::from(binary_path),
+            data_dir: PathBuf::from(data_dir),
+            network: network.to_owned(),
+            rpc_endpoint: rpc_endpoint.to_owned(),
+            num_connections,
+            log_file: PathBuf::from(log_file),
+            options: options
+                .iter()
+                .map(|v| (*v).to_owned())
+                .collect::<Vec<String>>(),
+        }
+    }
+}
+
+#[derive(Default)]
+struct ChildWrapper {
+    inner: Option<Child>,
+}
+
+impl ChildWrapper {
+    pub async fn kill(&mut self) -> anyhow::Result<()> {
+        if let Some(mut v) = self.inner.take() {
+            return Ok(v.kill()?);
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl AsyncDrop for ChildWrapper {
+    async fn async_drop(&mut self) {
+        let _ = self.kill().await;
+    }
+}
+
+#[derive(Default)]
+pub struct OctezNode {
+    inner: Arc<RwLock<AsyncDropper<ChildWrapper>>>,
```

Comment:
Hmm the inner child instance is never shared outside of OctezNode. Why does it need to be wrapped in `Arc<..>`?

Reply:
I'm also confused as to why it needs to be wrapped in the `RwLock<>` if the instance is never shared?

Reply:
At this moment it's indeed not very useful. I'm thinking about the case where one might want to clone the node, e.g.
```rs
let node = octez_node::OctezNode::spawn(...).await;
let copy = node.clone();
let trap = tokio::spawn(async {
    ...
    if (...) {
        node.kill().await;
    }
});
// continue with other logic
let _ = node.health_check().await;
```

Reply:
Ah ok